src/clustering/graph_based_clustering.pyx

```diff
@@ -0,0 +1,139 @@
+import networkx as nx
+import matplotlib.pyplot as plt
+from distance import ACGTContent
+from get_signature_metadata import get_metadata_for
+import numpy as np
+cimport numpy as np
+import time
+
+FLOATTYPE = np.float32
+INTTYPE = np.int
+ctypedef np.float32_t FLOATTYPE_t
+ctypedef np.int_t INTTYPE_t
+
+cdef class GraphBasedClustering(object):
+  """
+    A clustering implementation which keeps the VLMCs as nodes in a graph.
+    There are several methods for creating the clusters:
+    * Add edges where the distance between two nodes is less than some user-specified threshold, with
+    one of the distance functions.
+    * Add the minimum distance for every node which isn't in the same cluster already.
+  """
+
+  cdef double threshold
+  cdef list vlmcs
+  cdef d
+
+  def __init__(self, threshold, vlmcs, d):
+    self.threshold = threshold
+    self.vlmcs = vlmcs
+    self.d = d
+
+  cpdef cluster(self, clusters, min_distance=True, draw_graph=False):
+    G = nx.Graph()
+    G.add_nodes_from(self.vlmcs)
+    if min_distance:
+      self._cluster_with_min_distance(G, clusters)
+    else:
+      self._cluster_with_threshold(G)
+
+    if draw_graph:
+      self._draw_graph(G)
+    self._print_connected_components(G)
+
+  cdef _cluster_with_threshold(self, G):
+    for vlmc in self.vlmcs:
+      self._add_edges_from(G, vlmc)
+
+  cdef _add_edges_from(self, G, vlmc):
+    nearby_vlmcs = []
+    for other in self.vlmcs:
+      dist = self.d.distance(vlmc, other)
+      if (dist < self.threshold):
+        G.add_edge(vlmc, other, weight=dist)
+
+  cdef _cluster_with_min_distance(self, G, num_clusters):
+    start_time = time.time()
+
+    cdef np.ndarray[FLOATTYPE_t, ndim=2] distances = self._calculate_distances()
+
+    distance_time = time.time() - start_time
+    start_time = time.time()
+
+    cdef np.ndarray[FLOATTYPE_t, ndim=2] sorted_distances = distances[distances[:,2].argsort()]
+
+    sorting_time = time.time() - start_time
+    start_time = time.time()
+
+    clustering = {}
+    cdef FLOATTYPE_t i_t
+    for i, vlmc in enumerate(self.vlmcs):
+      i_t = i
+      clustering[i_t] = vlmc.name
+
+    connections_to_make = len(self.vlmcs) - num_clusters
+    for _ in range(connections_to_make):
+      # Add an edge for the shortest distance
+      # Take the smallest distance
+      [left, right, dist] = sorted_distances[0]
+      sorted_distances = np.delete(sorted_distances, 0, axis=0)
```

Comment:
Can we keep an index for the currently smallest distance instead of mutating the array